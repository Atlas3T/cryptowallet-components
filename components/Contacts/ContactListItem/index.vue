<template>
  <div
    class="contact-list-item"
    @click.prevent="clickHandler"
  >
    <div class="columns is-mobile no-margin-bottom">
      <div class="column is-2">
        <div class="circle" />
      </div>
      <div class="column">
        <div class="columns is-mobile no-margin-bottom">
          <div class="column is-7 no-padding-bottom">
            {{ contact.displayName }}
          </div>
        </div>
        <div class="wallet-address">
          <div v-if="contact.address">
            {{ contact.address }}
          </div>
          <div v-else>
            {{ $t('noWalletForContact') }}
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'ContactListItem',
  props: {
    contact: {
      type: Object,
      required: true,
    },
    clickItemAction: {
      type: String,
      required: false,
    },
  },
  methods: {
    /**
     * Initiates sending the invitation
     */
    clickHandler() {
      switch (this.clickItemAction) {
        case 'app-invitation':
          this.invitation = new this.AppInvitation({ vm: this, contact: this.contact });
          this.invitation.send();
          break;
        default:
          return false;
      }
      return false;
    },
  },
};
</script>

<style scoped>
.contact-list-item {
  margin: 1rem 0;
}

.wallet-address {
  font-size: 0.7rem;
}

.no-padding-bottom {
  padding-bottom: 0;
}

.no-margin-bottom {
  margin-bottom: 0;
}

.circle {
  background: #ececec;
  border-radius: 100%;
  height: 2.5rem;
  width: 2.5rem;
  margin-top: 0.1rem;
}
</style>
