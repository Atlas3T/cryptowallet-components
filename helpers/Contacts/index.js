import Vue from 'vue';

/**
 * Handles the contacts import and saves them in the store
 */
export default function Contacts() {
  this.vm = new Vue();
  this.import = (contacts) => {
    this.contacts = contacts;
    this.vm.$once('contactsImported', () => { return this.success(); });
    this.vm.$once('contactsImportFailure', () => { return this.failure(); });
    this.vm.$once('contactsImportError', (event) => { return this.error(event.payload); });

    try {
      const options = new ContactFindOptions();
      options.multiple = true;
      const fields = ['*'];
      navigator.contacts.find(fields, (contactsArray) => {
        this.contacts.contacts = contactsArray;
        this.vm.$emit('contactsImported');
      }, () => {
        this.vm.$emit('contactsImportFailure');
      }, options);
    } catch (err) {
      this.vm.$emit('contactsImportError', { payload: err });
    }

    return this;
  };

  this.on = (event, callback) => {
    this[event] = callback;
    return this;
  };
}
