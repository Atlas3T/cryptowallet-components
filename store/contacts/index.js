export default {
  namespaced: true,
  state: {
    contacts: [
      {
        id: 1,
        displayName: 'Terry Tibbs',
        address: 'ncousohfodicwjdclwdnc',
      },
    ],
  },
};
