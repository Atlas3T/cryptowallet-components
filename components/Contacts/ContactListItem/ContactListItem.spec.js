import { shallowMount } from '@vue/test-utils';
import ContactListItem from '@/components/Contacts/ContactListItem';
import { localVue, i18n, createRouter } from '@/helpers/SetupLocalVue';
import { createMocks as createStoreMocks } from '@/store/__mocks__/store.js';
import AppInvitationMock from '@/boot/AppInvitation/mock/';

describe('ContactListItem.vue', () => {
  let storeMocks;
  let wrapper;
  let store;
  let router;

  const propsData = {
    contact: {
      id: 1,
      displayName: 'Fio',
      address: '1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE',
    }
  };

  function wrapperInit (options) {
    return shallowMount(ContactListItem, options);
  }

  function storeInit (custom) {
    storeMocks = createStoreMocks(custom);
    router = createRouter(storeMocks.store);
    router.push({ path: `/` });
    wrapper = wrapperInit({ i18n, router, localVue, store: storeMocks.store, propsData });
    store = wrapper.vm.$store;
  }

  beforeEach(() => storeInit());

  it('renders and matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders a div with .contact-list-item class', () => {
    expect(wrapper.contains('div.contact-list-item')).toBe(true);
  });

  it('renders contacts displayName', () => {
    expect(wrapper.html().includes(propsData.contact.displayName)).toBe(true);
  });

  it('renders a wallet address if it has the data', () => {
    expect(propsData.contact.address.length > 0).toBe(true);
    expect(wrapper.html().includes(propsData.contact.address)).toBe(true);
  });

  it('renders a custom message if there is no wallet address available', () => {
    const contactCopy = Object.assign({}, propsData.contact);
    delete contactCopy.address;
    wrapper.setProps({ contact: contactCopy });
    expect(wrapper.html().includes('No wallet address is available for this contact')).toBe(true);
  });

  it('it calls clickHandler() method when clicked on', () => {
    const clickHandlerMock = jest.fn();
    wrapper.setMethods({ clickHandler: clickHandlerMock });
    wrapper.find('.contact-list-item').trigger('click');
    expect(clickHandlerMock).toHaveBeenCalled();
  });

  it('creates an AppInvitation when clickItemAction prop === \'app-invitation\'', () => {
    wrapper.setProps({ clickItemAction: 'app-invitation' });
    wrapper.vm.AppInvitation = AppInvitationMock;
    wrapper.vm.clickHandler();;
    expect(wrapper.vm.invitation.send).toHaveBeenCalled();
  });

  it('returns false from clickHandler() method', () => {
    expect(wrapper.vm.clickHandler()).toBe(false);
  });
});
