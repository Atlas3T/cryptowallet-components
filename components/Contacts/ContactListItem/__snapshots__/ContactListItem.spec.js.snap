// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactListItem.vue renders and matches snapshot 1`] = `
<div
  class="contact-list-item"
>
  <div
    class="columns is-mobile no-margin-bottom"
  >
    <div
      class="column is-2"
    >
      <div
        class="circle"
      />
    </div>
     
    <div
      class="column"
    >
      <div
        class="columns is-mobile no-margin-bottom"
      >
        <div
          class="column is-7 no-padding-bottom"
        >
          Fio
        </div>
      </div>
       
      <div
        class="wallet-address"
      >
        <div>
          
          1CK6KHY6MHgYvmRQ4PAafKYDrg1ejbH1cE
        
        </div>
      </div>
    </div>
  </div>
</div>
`;
